File: src/targets/databases/cassandra/spar_idp_count.py
"""Counts SAML identity providers configured in the Cassandra spar keyspace.

A zero count means SAML SSO is not configured or all IdPs have been
deleted. This is significant when brig's optSettings.setSSOEnabled=true
because users will not be able to log in via SSO. See JCT-164.
"""

from __future__ import annotations

# Ours
from src.lib.base_target import BaseTarget
from src.lib.cql_types import CqlResult


class CassandraSparIdpCount(BaseTarget):
    """Counts configured SAML identity providers in the spar Cassandra keyspace.

    Queries spar.idp table (or spar.user for older schemas) and
    returns how many identity providers are configured. Zero means
    SAML SSO isn't set up, even if brig's SSO flag is enabled.
    """

    # Uses SSH to reach Cassandra nodes
    requires_ssh: bool = True

    @property
    def description(self) -> str:
        """Human-readable description of what this target checks."""
        return "Cassandra spar SAML identity provider count"

    @property
    def explanation(self) -> str:
        """Why this check exists and what determines healthy vs unhealthy."""
        return (
            "If brig's setSSOEnabled is true but no IdPs are registered in "
            "spar.idp, users can't log in via SSO. "
            "A missing spar.issuer_idp row is the usual cause (JCT-164)."
        )

    @property
    def unit(self) -> str:
        """Unit label shown next to the collected value."""
        return "IdPs"

    def collect(self) -> int:
        """Count SAML identity providers in the spar Cassandra keyspace.

        Returns:
            Number of rows in the spar.idp table (count of configured IdPs).

        Raises:
            RuntimeError: If the Cassandra query cannot be executed.
        """
        self.terminal.step("Counting SAML IdPs in Cassandra spar keyspace...")

        idp_count: int | None = self._count_via_cql()

        if idp_count is None:
            idp_count = self._count_via_cqlsh()

        if idp_count is None:
            raise RuntimeError(
                "Couldn't count spar.idp rows - native CQL and cqlsh both failed"
            )

        if idp_count == 0:
            self._health_info = (
                "No SAML IdPs registered in spar.idp. "
                "Severity depends on brig optSettings.setSSOEnabled - "
                "checked by the UI layer against kubernetes/configmaps/brig."
            )
        else:
            self._health_info = f"{idp_count} SAML IdP(s) configured in spar keyspace"

        return idp_count

    def _count_via_cql(self) -> int | None:
        """Count spar.idp rows via native CQL protocol.

        Returns:
            Row count, or None if the query failed.
        """
        try:
            result: CqlResult = self.run_cql_query("SELECT COUNT(*) FROM spar.idp")

            if result.rows:
                count_val: object = result.rows[0][0]
                return int(count_val) if count_val is not None else 0

            return 0
        except Exception as exc:
            self.terminal.step(f"Native CQL failed ({exc}), falling back to cqlsh...")
            return None

    def _count_via_cqlsh(self) -> int | None:
        """Count spar.idp rows via cqlsh as fallback.

        Returns:
            Row count, or None if cqlsh also failed.
        """
        result = self.run_cqlsh("SELECT COUNT(*) FROM spar.idp")
        output: str = result.stdout.strip()

        if not output:
            return None

        # cqlsh tabular output has separator lines like "------"
        # The data rows sit between the first and second separators.
        # Only parse integers from that region to avoid picking up
        # stray numbers from headers, footers, or timing lines.
        lines: list[str] = output.split("\n")
        in_data_region: bool = False

        for line in lines:
            stripped: str = line.strip()

            # Detect separator lines (e.g. "------" or "---+---")
            if stripped and all(ch in "-+" for ch in stripped):
                if in_data_region:
                    # Second separator ends the data region
                    break
                in_data_region = True
                continue

            if in_data_region:
                try:
                    value: int = int(stripped)
                    return value
                except ValueError:
                    continue

        return None

File: ui/src/checkers/cassandra/spar_idp_count.ts
/**
 * Reports how many SAML identity providers are configured in Cassandra.
 *
 * Pulls the IdP count from Cassandra and checks brig's config to see if
 * SSO is actually enabled. Zero IdPs is only a problem if brig has
 * setSSOEnabled set to true, otherwise it's expected. This catches the
 * JCT-164 situation where spar.issuer_idp is empty even though SSO is
 * turned on in brig.
 */

// External
import yaml from 'js-yaml'

// Ours
import { BaseChecker, type CheckResult } from '../base_checker'
import { parse_number, type DataLookup } from '../data_lookup'

export class CassandraSparIdpCountChecker extends BaseChecker {
    readonly path: string = 'cassandra/spar_idp_count'
    readonly name: string = 'SAML IdP count in Cassandra spar (see: JCT-164)'
    readonly category: string = 'Data / Cassandra'
    readonly interest = 'Health, Setup' as const

    readonly requires_ssh: boolean = true
    readonly explanation: string = 'Counts **SAML identity providers** configured in Cassandra and cross-references with brig SSO settings. When SSO is enabled but no IdPs exist, users cannot authenticate via SSO and are locked out of Wire.'

    check(data: DataLookup): CheckResult {
        // Auto-detect SSO status from the galley feature flag rather than
        // relying on a user toggle. If galley's SSO flag is not "enabled",
        // this check is not applicable.
        const sso_flag_point = data.get('config/galley_sso_flag')
        if (sso_flag_point?.value) {
            const sso_flag: string = String(sso_flag_point.value).toLowerCase()
            if (!sso_flag.includes('enabled')) {
                return {
                    status: 'not_applicable',
                    status_reason: 'SSO is not enabled in galley (auto-detected). SAML IdP count check does not apply.',
                    display_value: 'skipped (SSO disabled)',
                }
            }
        } else if (data.config && !data.config.options.expect_sso) {
            // Fall back to the user toggle if auto-detection data is unavailable
            return {
                status: 'not_applicable',
                status_reason: 'SSO is not enabled — SAML IdP count check does not apply.',
                display_value: 'skipped',
            }
        }

        const point = data.get('databases/cassandra/spar_idp_count')

        // Couldn't collect the data in the first place
        if (!point) {
            return {
                status: 'gather_failure',
                status_reason: 'SAML IdP count data was not collected from Cassandra spar.',
                fix_hint: '1. Verify SSH connectivity to the Cassandra nodes\n2. Check that `cqlsh` can query the spar keyspace: `cqlsh -e "SELECT COUNT(*) FROM spar.idp"`\n3. Review the gatherer logs for connection errors or authentication failures',
                recommendation: 'Couldn\'t collect the IdP count from Cassandra spar.',
            }
        }

        const count = parse_number(point)

        // Value wasn't parseable as a number
        if (count === null) {
            return {
                status: 'gather_failure',
                status_reason: 'SAML IdP count value from Cassandra spar could not be parsed as a number.',
                recommendation: 'The IdP count from Cassandra spar was not a valid number. Check the raw gatherer output.',
                raw_output: point.raw_output,
            }
        }

        // At least one IdP is set up, so SSO can work
        if (count > 0) {
            return {
                status: 'healthy',
                status_reason: 'Found **{{idp_count}}** SAML IdP(s) configured in Cassandra spar, SSO authentication is available.',
                display_value: count,
                display_unit: 'IdPs',
                raw_output: point.raw_output,
                template_data: { idp_count: count },
            }
        }

        // No IdPs, so we need to check if SSO is actually enabled
        const sso_enabled: boolean | null = this._read_brig_sso_enabled(data)

        if (sso_enabled === true) {
            // Bad news: SSO is on but no IdPs configured, so users are locked out
            return {
                status: 'unhealthy',
                status_reason: 'No SAML IdPs are configured in Cassandra spar, but SSO is **enabled** in brig config, so users cannot authenticate via SSO.',
                fix_hint: '1. Verify the spar IdP table: `cqlsh -e "SELECT * FROM spar.idp"`\n2. Check if the `spar.issuer_idp` table is also empty: `cqlsh -e "SELECT * FROM spar.issuer_idp"`\n3. If IdPs were previously configured, check for data loss or migration issues\n4. To add an IdP, use the Wire admin API or re-run the SAML SSO setup\n5. Alternatively, if SSO is not needed, disable it in brig config: set `setSSOEnabled: false` in the brig ConfigMap',
                recommendation: 'No IdPs in Cassandra spar.idp, but brig has setSSOEnabled turned on. That means users can\'t log in via SSO. Either add an IdP or disable SSO in brig config.',
                display_value: 0,
                display_unit: 'IdPs',
                raw_output: point.raw_output,
            }
        }

        if (sso_enabled === false) {
            // SSO is off, so having zero IdPs is fine
            return {
                status: 'healthy',
                status_reason: 'No SAML IdPs configured, but SSO is **disabled** in brig config, so none are needed.',
                display_value: 0,
                display_unit: 'IdPs',
                raw_output: point.raw_output,
            }
        }

        // Can't reach brig config or the flag isn't there, so we're not sure
        return {
            status: 'warning',
            status_reason: 'No SAML IdPs configured in Cassandra spar, and brig config could not be read to determine if SSO is enabled.',
            fix_hint: '1. Manually check the spar IdP table: `cqlsh -e "SELECT * FROM spar.idp"`\n2. Check brig ConfigMap for `setSSOEnabled`: `kubectl get configmap brig -o yaml | grep setSSOEnabled`\n3. If SSO is enabled and no IdPs exist, users cannot log in via SSO\n4. Either add an IdP via the Wire admin API or disable SSO in brig config',
            recommendation: 'No IdPs in Cassandra spar.idp. Couldn\'t read brig config to check if SSO is enabled. If it is, users won\'t be able to log in via SSO.',
            display_value: 0,
            display_unit: 'IdPs',
            raw_output: point.raw_output,
        }
    }

    /**
     * Pull the setSSOEnabled flag from brig's ConfigMap.
     *
     * @param data all the collected data
     * @returns true or false if we can read it, null if we can't
     */
    private _read_brig_sso_enabled(data: DataLookup): boolean | null {
        const brig_point = data.get('kubernetes/configmaps/brig')

        if (!brig_point?.value) {
            return null
        }

        let parsed: Record<string, unknown>

        try {
            parsed = yaml.load(String(brig_point.value)) as Record<string, unknown>
        } catch {
            return null
        }

        if (!parsed || typeof parsed !== 'object') {
            return null
        }

        const opt_settings = parsed.optSettings as Record<string, unknown> | undefined
        const sso_flag = opt_settings?.setSSOEnabled

        if (typeof sso_flag !== 'boolean') {
            return null
        }

        return sso_flag
    }
}

export default CassandraSparIdpCountChecker
